### PR TITLE
daemon: fix clang and kernel versions output

### PR DESCRIPTION
**Pull request by aanm**
Previously we were only priting the major.minor versions of clang and
the kernel. This commit will try to retrieve the kernel and clang patch
versions. On a worst case it will print the patch version "0"

Previously:
clang (3.8.0) and kernel (4.9.0) versions: OK!
Now:
clang (3.8.1) and kernel (4.9.17) versions: OK!

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Printing patch clang and kernel patch versions when starting cilium.
```

It is a fair assumption the kernel will have `major.minor.patch[-random-string]`?

### CODE REVIEW

File: daemon/main.go [hunk 1]
```diff
@@ -160,10 +160,19 @@ func getKernelVersion() (*go_version.Version, error) {
...
 		return nil, fmt.Errorf("unable to get kernel version from %q", string(verOut))
 	}
-	return go_version.NewVersion(strings.Join(verStrs[:2], "."))
+	// We are assuming the kernel version will be something as:
+	// 4.9.17-040917-generic
+
+	// If verStrs is []string{ "4", "9", "17-040917-generic" }
+	// then we need to split the patch string to get a number.
+	patchVersion := strings.Split(verStrs[2], "-")
+	if len(patchVersion) != 0 {
```
Comment (raybejjani) [inline daemon/main.go]: Do we need to clear the third element of verStr when len(patchVersion) == 0? Otherwise we would just recombine the version including the full patch string.

File: daemon/main.go [hunk 2]
```diff
@@ -178,10 +187,10 @@ func getClangVersion(filePath string) (*go_version.Version, error) {
 	}
 	// at this point res is []string{"clang", "version", "maj.min.patch"}
 	verStrs := strings.Split(res[2], ".")
-	if len(verStrs) < 2 {
-		return nil, fmt.Errorf("unable to get kernel version from %q", string(verOut))
+	if len(verStrs) < 3 {
+		return nil, fmt.Errorf("unable to get clang version from %q", string(verOut))
 	}
-	return go_version.NewVersion(strings.Join(verStrs[:2], "."))
+	return go_version.NewVersion(strings.Join(verStrs[:3], "."))
```
Comment (raybejjani) [inline daemon/main.go:202]: I realise it was like this before but what's the point here, that we verify the version has 3 components? If so, wouldn't it be shorter to write the full regex to match "clang version 1.2.3" and error if there is no match, and otherwise just return the 1.2.3 bit?

File: daemon/main.go [hunk 3]
```diff
@@ -160,10 +160,19 @@ func getKernelVersion() (*go_version.Version, error) {
...
-	if len(verStrs) < 2 {
+	if len(verStrs) < 3 {
 		return nil, fmt.Errorf("unable to get kernel version from %q", string(verOut))
 	}
-	return go_version.NewVersion(strings.Join(verStrs[:2], "."))
+	// We are assuming the kernel version will be something as:
+	// 4.9.17-040917-generic
+
+	// If verStrs is []string{ "4", "9", "17-040917-generic" }
+	// then we need to split the patch string to get a number.
```
Comment (borkmann) [inline daemon/main.go]: In my case it looks like:

`$ uname -a
Linux localhost.localdomain 4.14.0-rc7+ #187 SMP Thu Nov 23 16:26:44 CET 2017 x86_64 x86_64 x86_64 GNU/Linux`

Couple of more possible examples:
```
4.10.0
4.12.0+
4.12.8
4.14.0-rc7+
```
Would they get detected as well?

**Review by raybejjani — APPROVED** [review]
I'm not sure I follow how it would print 0 in the worst case (with the kernel version I think it prints the patch number as is) but it's generally harmless.

File: daemon/main.go (hunk 2 above)
Comment (aanm) [reply daemon/main.go:202]: I didn't get it, can you rephrase?

File: daemon/main.go (hunk 1 above)
Comment (aanm) [reply daemon/main.go]: I've rewrote this function again

File: daemon/main.go (hunk 3 above)
Comment (aanm) [reply daemon/main.go]: I've added unit tests
```
                {"4.10.0", mustHaveVersion("4.10.0")},
		{"4.10", mustHaveVersion("4.10.0")},
		{"4.12.0+", mustHaveVersion("4.12.0")},
		{"4.12.8", mustHaveVersion("4.12.8")},
		{"4.14.0-rc7+", mustHaveVersion("4.14.0")},
		{"4.9.17-040917-generic", mustHaveVersion("4.9.17")},
		{"4.9.generic", mustHaveVersion("4.9.0")},
```
